### ISSUE TITLE
Files generated during precompile do not get added to compilation

### ISSUE THREAD

**Issue by pranavkm** [dotnet/cli#3807]
## Steps to reproduce

Consider a project.json with a scripts node like so

```
"scripts": {
    "precompile": "cmd /c echo namespace Test { public class Foo {} } > Test.cs"
  }
```

Referencing `Test.Foo` does not work in the first compile when the file is not on disk. Having an empty Test.cs which gets populated during `precompile` works. It looks like the build is calculating the set of files to compile earlier than precompile. Perhaps it needs to do it after?

`dotnet --info` output: `1.0.0-preview3-003150`

**Comment by ocdi** [comment]
I am trying to use the precompile script to generate some files and have the same issue, the first compile they aren't included. I'm using 1.0.0-preview2-003121.

**Comment by stickybun** [comment]
This is also an issue for embedding files which are created during precompile.

For example:

**project.json**
```
"buildOptions": {
    "embed": {
      "include": "*.txt"
    }
  }
```
```
"scripts": {
    "precompile": [
      "bash -c \"touch $(date +%H%M%S).txt\""
    ]
  }
```
or on Windows:
```
  "scripts": {
    "precompile": [
      "cmd /c echo \"\" >>  %time:~1,1%%time:~3,2%%time:~6,2%.txt"
    ]
  }
```

**Program.cs**
```
public static void Main(string[] args)
{
    var embeddedCount = System.Reflection.Assembly.GetEntryAssembly().GetManifestResourceNames().Count(m=>m.EndsWith(".txt"));
    var actualCount = System.IO.Directory.GetFiles(".","*.txt").Length;
    Console.WriteLine($"Number of txt resources: {embeddedCount}, actual number of files: {actualCount}");
}
```

**Output**
`Number of txt resources: 0, actual number of files: 1`
`Number of txt resources: 1, actual number of files: 2`
`Number of txt resources: 2, actual number of files: 3`

**Expected Output**
`Number of txt resources: 1, actual number of files: 1`

_Version:            1.0.0-preview2-003121_

**Comment by TanukiSharp** [comment]
I confirm the issue, with `1.0.0-preview2-003121`.

It looks like the code that runs precompile actions does not wait for those actions to be done before starting the compilation.

**Comment by chwarr** [comment]
This looks like a duplicate of https://github.com/dotnet/cli/issues/1475 to me. That issue mentions a possible workaround: have dummy files that get overwritten by the precompile scripts. This is working for me.

I doubt that this will get fixed, with the planned move to MSBuild-based builds.

(Though I work for Microsoft, I'm not associated with this project.)

**Comment by livarcocc** [comment]
This will work with preview3.
